src/gt4py/next/common.py

```diff
@@ -117,10 +129,102 @@ def __getitem__(self, index: int | slice) -> int | UnitRange:
             else:
                 raise IndexError("UnitRange index out of range")
 
+    def __and__(self, other: UnitRange) -> UnitRange:
+        start = max(self.start, other.start)
+        stop = min(self.stop, other.stop)
+        return UnitRange(start, stop)
```

Comment:
As mypy told me, this violates Liskov substitution principle, we need to cover the case where other is a Set[int], e.g something like
```suggestion
    def __and__(self, other: Set[int]) -> UnitRange:
        if isinstance(other, UnitRange):
            start = max(self.start, other.start)
            stop = min(self.stop, other.stop)
            return UnitRange(start, stop)
        else:
            other_start = min(other)
            other_stop = max(other)
            if len(other) != other_stop - other_start:
                raise ValueError("Intersection with non-contiguous range is not supported.")
            start = max(self.start, other_start)
            stop = min(self.stop, other_stop)
            return UnitRange(start, stop)
```

Reply:
Do we want to support intersections with sets too then or only UnitRange?

Reply:
this will solve the problem that you had with adding `Set` to the inheritance

Reply:
alternatively we could also just `raise NotImplementedError()` for the case where other is not `UnitRange`